pkg/collector/runner/num_workers_test.go

```diff
@@ -0,0 +1,227 @@
+		time.Sleep(time.Millisecond * 100)
+
+	case pythonCheck:
+		runPythonCheck()
+	}
+
+	nc.hasRun = true
+	return nil
+}
+
+// Efficiency test for the number of check workers running
+// Note: use the -v flag when testing to see the output
+func TestUpdateNumWorkers(t *testing.T) {
+	if !testingEfficiency {
+		return
+	}
+
+	t.Logf("Testing efficiency on a machine with %v cores", runtime.NumCPU())
+	if checkType == pythonCheck {
+		// Initialize the python interpreter
+		state := py.Initialize(".", "../dist")
+		defer python.PyEval_RestoreThread(state)
+	}
+
+	// Run the time tests
+	interval := false
+	for i := 0; i < 2; i++ {
+		if interval {
+			t.Logf("********* Starting the time efficiency test (%v repeats) *********", numIntervals)
+		} else {
+			t.Log("********* Starting the time efficiency test (single) *********")
+		}
+
+		checksToRun := [10]int{5, 15, 25, 35, 45, 55, 65, 75, 85, 100}
+
+		for _, n := range checksToRun {
+			ti := timeToComplete(n, interval)
+
+			if prettyOutput {
+				t.Logf("Time to run %v checks: %v", n, ti.Seconds())
+			} else {
+				t.Logf("%v", ti.Seconds())
+			}
+		}
+
+		if numIntervals == 0 {
+			break
+		}
+		interval = true
+	}
+
+	if !memoryTest {
+		return
+	}
+	// Run the memory test
+	r := NewRunner()
+	curr, _ := strconv.Atoi(runnerStats.Get("Workers").String())
+	runnerStats.Add("Workers", int64(curr*-1))
+	m := &runtime.MemStats{}
+
+	t.Log("********* Starting memory test *********")
+	runtime.ReadMemStats(m)
+
+	if prettyOutput {
+		t.Logf("At start:")
+		t.Logf("\tAlloc = %v\tSys = %v\tHeapAlloc = %v\tHeapSys = %v\tHeapObj = %v\t",
+			m.Alloc/1024, m.Sys/1024, m.HeapAlloc, m.HeapSys, m.HeapObjects)
+	} else {
+		t.Logf("%v\t%v\t%v\t%v\t%v\t", m.Alloc/1024, m.Sys/1024, m.HeapAlloc, m.HeapSys, m.HeapObjects)
+	}
+
+	for i := 1; i < 500; i++ {
+		c := NumWorkersCheck{name: "Check" + strconv.Itoa(i)}
+		if !static {
+			r.UpdateNumWorkers(int64(i))
+		}
+		r.pending <- &c
+
+		if i%100 == 0 {
+			runtime.ReadMemStats(m)
+
+			if prettyOutput {
+				t.Logf("After %d checks:", i)
+				t.Logf("\tAlloc = %v\tSys = %v\tHeapAlloc = %v\tHeapSys = %v\tHeapObj = %v\t",
+					m.Alloc/1024, m.Sys/1024, m.HeapAlloc, m.HeapSys, m.HeapObjects)
+			} else {
+				t.Logf("%v\t%v\t%v\t%v\t%v\t", m.Alloc/1024, m.Sys/1024, m.HeapAlloc, m.HeapSys, m.HeapObjects)
+			}
+		}
+	}
+}
+
+func timeToComplete(numChecks int, runMultiple bool) time.Duration {
+	r := NewRunner()
+
+	// Reset the stats to having 4 workers
+	curr, _ := strconv.Atoi(runnerStats.Get("Workers").String())
+	runnerStats.Add("Workers", int64(curr*-1)+4)
+
+	start := time.Now()
+
+	// Initialize the correct number of checks in the channel
+	for i := 1; i < numChecks; i++ {
+		c := NumWorkersCheck{name: "Check" + strconv.Itoa(i)}
+		if !static {
+			r.UpdateNumWorkers(int64(i))
+		}
+		r.pending <- &c
+	}
+
+	if runMultiple {
+		// To imitate a check running at an interval (UpdateNumWorkers doesn't run again)
+		for j := 0; j < numIntervals; j++ {
```

Comment:
Why don't we just pass `numIntervals` as the parameter to `timeToComplete()` and drop `runMultiple`? I think it would make for the more natural interface. Otherwise we have this duplicated code, and nested code in this if, etc....